fuel-merkle/src/sparse/proof.rs

```diff
@@ -0,0 +1,692 @@
+use crate::{
+    common::{
+        path::{
+            Instruction,
+            Path,
+        },
+        Bytes32,
+        Prefix,
+        ProofSet,
+    },
+    sparse::Node,
+};
+
+use alloc::vec::Vec;
+use core::{
+    fmt,
+    fmt::Debug,
+};
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Proof {
+    Inclusion(InclusionProof),
+    Exclusion(ExclusionProof),
+}
+
+impl Proof {
+    pub fn root(&self) -> &Bytes32 {
+        match self {
+            Proof::Inclusion(proof) => &proof.root,
+            Proof::Exclusion(proof) => &proof.root,
+        }
+    }
+
+    pub fn proof_set(&self) -> &ProofSet {
+        match self {
+            Proof::Inclusion(proof) => &proof.proof_set,
+            Proof::Exclusion(proof) => &proof.proof_set,
+        }
+    }
+
+    pub fn is_inclusion(&self) -> bool {
+        match self {
+            Proof::Inclusion(_) => true,
+            Proof::Exclusion(_) => false,
+        }
+    }
+
+    pub fn is_exclusion(&self) -> bool {
+        !self.is_inclusion()
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct InclusionProof {
+    pub root: Bytes32,
+    pub proof_set: ProofSet,
+}
+
+impl InclusionProof {
+    pub fn verify<K: Into<Bytes32>, V: AsRef<[u8]>>(&self, key: K, value: &V) -> bool {
+        let Self { root, proof_set } = self;
+        let key = key.into();
+        let leaf = Node::create_leaf(&key, value);
+        let mut current = *leaf.hash();
+        for (i, side_hash) in proof_set.iter().enumerate() {
+            let index = u32::try_from(proof_set.len() - 1 - i).expect("Index is valid");
+            let prefix = Prefix::Node;
+            current = match key.get_instruction(index).expect("Infallible") {
+                Instruction::Left => Node::calculate_hash(&prefix, &current, side_hash),
+                Instruction::Right => Node::calculate_hash(&prefix, side_hash, &current),
+            };
+        }
+        current == *root
+    }
+}
+
+impl Debug for InclusionProof {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let proof_set = self.proof_set.iter().map(hex::encode).collect::<Vec<_>>();
+        f.debug_struct("InclusionProof")
+            .field("Root", &hex::encode(self.root))
+            .field("Proof set", &proof_set)
+            .finish()
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct ExclusionProof {
+    pub root: Bytes32,
+    pub proof_set: ProofSet,
+    pub(crate) leaf: Node,
+}
+
+impl ExclusionProof {
+    pub fn verify<K: Into<Bytes32>>(&self, key: K) -> bool {
+        let Self {
+            root,
+            proof_set,
+            leaf,
+        } = self;
+        let key = key.into();
+        let mut current = *leaf.hash();
```

Comment:
Based on how the proof was generated, the leaf can be either a real leaf or a placeholder. In the case of a placeholder, the hash will be the zero sum. In the case of a real leaf, this leaf corresponds to the "closest" leaf available at the time of generation, and the hash will correspond to that leaf.

Reply:
Almost fixed=D The problem is that getter `leaf_key` has an internal assert. So if someone provides a `node,` it will fail.

Added a test case with a panic=)

I think if the node is not a placeholder or a leaf, we need to return false. Overwise the proof provider may try to trick us=)